### PR TITLE
server/auth: fix panic a malformed jwt generation and add test-cases

### PR DESCRIPTION
**Pull request by ArkaSaha30**
This PR is to fix panic on identical JWT token generation and authentication as well as add test cases for the same.

Extension of #15002

Fixes: https://github.com/etcd-io/etcd/issues/14931

### CODE REVIEW

**Comment by ahrtr** [comment]
Thanks @ArkaSaha30 . Please take this as a high priority, because it's planned to be included in 3.5.8.

**Comment by ArkaSaha30** [comment]
Hi @ahrtr , just wanted to confirm if I need to create `public.pem` `private.pem` in `./tests/fixtures/` as a part of adding the test case?

**Comment by ahrtr** [comment]
> Hi @ahrtr , just wanted to confirm if I need to create `public.pem` `private.pem` in `./tests/fixtures/` as a part of adding the test case?

Can  you reuse the existing certificate and key files? Please refer to `server/auth/jwt_test.go`.

**Comment by ArkaSaha30** [comment]
In `server/auth/jwt_test.go` will it be an addition to the map object in `TestJWTInfo()`. I am not sure what exact unit test  case should be written around the change

**Comment by ahrtr** [comment]
You need to intentionally generate a JWT token without `username` or `revision` field.

**Comment by ahrtr** [comment]
@ArkaSaha30 Please remove `WIP` from the tile once you finish this PR, and then I will take a look at this PR soon.

File: server/auth/jwt_test.go [hunk 1]
```diff
@@ -132,6 +132,35 @@ func testJWTInfo(t *testing.T, opts map[string]string) {
...
+func testJWTNoUsernameRevisionInfo(t *testing.T, opts map[string]string, flag string) {
+	lg := zap.NewNop()
+	jwt, err := newTokenProviderJWT(lg, opts)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ctx := context.TODO()
+
+	ai, ok := jwt.info(ctx, "aaa", 120)
```
Comment (ahrtr) [inline server/auth/jwt_test.go:158]: `aaa` isn't a valid JWT token at all. You need to generate a valid JWT token, but without `username` or `revision` field.

Comment (ahrtr) [reply server/auth/jwt_test.go:158]: Please refer to https://github.com/etcd-io/etcd/blob/5327d91c20935a4324ace7aefa4a703c1e0d76a2/server/auth/jwt.go#L89-L105

For example to generate a JWT token without `revision`
```
	tk := jwt.NewWithClaims(t.signMethod,
		jwt.MapClaims{
			"username": username,
			"exp":      time.Now().Add(t.ttl).Unix(),
		})

	token, err := tk.SignedString(t.key)
	if err != nil {
		t.lg.Debug(
			"failed to sign a JWT token",
			zap.String("user-name", username),
			zap.Uint64("revision", revision),
			zap.Error(err),
		)
		return "", err
	}
```

Comment (ArkaSaha30) [reply server/auth/jwt_test.go]: Generated token without `username` and `revision` and used that in the test function

File: server/auth/jwt_test.go [hunk 2]
```diff
@@ -132,6 +134,68 @@ func testJWTInfo(t *testing.T, opts map[string]string) {
...
+func TestJWTNoUsernameRevision(t *testing.T) {
+	optsMap := map[string]map[string]string{
+		"RSA": {
+			"pub-key":     jwtRSAPubKey,
+			"priv-key":    jwtRSAPrivKey,
+			"sign-method": "RS256",
+		},
+	}
+
+	t.Run("RSA", func(tt *testing.T) {
```
Comment (ahrtr) [inline server/auth/jwt_test.go:146]: Please use table driven golang sub test to test the following two cases:
1. Only  has username and exp;
2. Only has revision and exp;

File: server/auth/jwt_test.go [hunk 3]
```diff
@@ -132,6 +134,68 @@ func testJWTInfo(t *testing.T, opts map[string]string) {
...
+		t.Fatalf("%#v", aerr)
+	}
+
+	ai, ok := jwt.info(ctx, token, 120)
+	if ok || ai != nil {
+		t.Fatalf("expected empty token to fail to authenticate, got %+v", ai)
+	}
+}
+
+func (t *tokenJWT) testJWTNoUsernameRevisionAssign(ctx context.Context) (string, error) {
```
Comment (ahrtr) [inline server/auth/jwt_test.go:171]: Please change it to a function: remove the receiver.

File: server/auth/jwt_test.go (hunk 2 above)
Comment (ahrtr) [reply server/auth/jwt_test.go:146]: Please refer to example https://github.com/etcd-io/etcd/blob/caa775d23563efdec9b4fa6519fdfbfe30985e5f/server/etcdserver/server_test.go#L67

**Comment by serathius** [comment]
This PR is blocking v3.5.8 release, please provide fixes as fast you can. Plan is to release next week, so please provide fixes till Monday.

**Comment by ahrtr** [comment]
> This PR is blocking v3.5.8 release, please provide fixes as fast you can. Plan is to release next week, so please provide fixes till Monday or we will need to finish the PR. Sorry for not making the importance the PR earlier.

I will work with @ArkaSaha30 to get it done next Monday. See also https://github.com/etcd-io/etcd/issues/15585#issuecomment-1500689116

**Comment by ArkaSaha30** [comment]
I'm already working on it, I will push the changes for review ASAP.

**Comment by ahrtr** [comment]
@ArkaSaha30 Please follow https://github.com/etcd-io/etcd/pull/15676

**Comment by ahrtr** [comment]
Usually we fix issue on main branch firstly, and then backport to 3.5 and 3.4. But we have no time to wait, because we need to release both 3.5.8 and 3.4.25 soon to resolve some CVEs.  So I delivered two PRs myself for both 3.5 and 3.4 before this one is approved/merged. Sorry for the confusion. 

Anyway, please follow the same way to update this PR.

**Comment by ArkaSaha30** [comment]
Thank you @ahrtr, I am updating this PR accordingly.

**Review by jmhbnz — APPROVED** [review]
LGTM. Thanks @ArkaSaha30.

**Comment by ahrtr** [comment]
minor comment: please consider to squash the second and third commits.

**Review by ahrtr — APPROVED** [review]
LGTM

**Comment by ahrtr** [comment]
cc @mitake @spzala PTAL, this should be safe change.

**Review by serathius — APPROVED** [review]

**Comment by mitake** [comment]
LGTM, thanks @ArkaSaha30 !

**Pull request merged**